domain/user/state/state.go

```diff
@@ -0,0 +1,358 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+	domainuser "github.com/juju/juju/domain/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+	databaseutils "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func addUser(ctx context.Context, tx *sql.Tx, uuid domainuser.UUID, user user.User, creatorUUID domainuser.UUID) error {
```

Comment:
Put the private methods at the end of the file, with the exported methods in the `State` type listed first.